wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileManager.java

```diff
@@ -121,23 +187,85 @@ public void resolveReference(DevfileImpl devfile, FileContentProvider fileConten
     }
   }
 
-  private DevfileImpl parse(String content, ValidationFunction validationFunction)
+  private DevfileImpl parse(
+      String content, ObjectMapper mapper, Map<String, String> overrideProperties)
       throws DevfileFormatException {
-    JsonNode parsed = validationFunction.validate(content);
-
     DevfileImpl devfile;
     try {
-      devfile = objectMapper.treeToValue(parsed, DevfileImpl.class);
+      JsonNode parsed = mapper.readTree(content);
+      parsed = applyPropertiesOverride(parsed, overrideProperties);
+      schemaValidator.validate(parsed);
+      devfile = mapper.treeToValue(parsed, DevfileImpl.class);
     } catch (JsonProcessingException e) {
       throw new DevfileFormatException(e.getMessage());
+    } catch (IOException e) {
+      throw new DevfileFormatException("Unable to parse Devfile. Error: " + e.getMessage());
     }
-
     integrityValidator.validateDevfile(devfile);
     return devfile;
   }
 
-  @FunctionalInterface
-  private interface ValidationFunction {
-    JsonNode validate(String content) throws DevfileFormatException;
+  private JsonNode applyPropertiesOverride(
+      JsonNode devfileNode, Map<String, String> overrideProperties) throws DevfileFormatException {
+    for (Map.Entry<String, String> entry : overrideProperties.entrySet()) {
+      // prepare stuff
+      String[] pathSegments = entry.getKey().split("\\.");
+      if (pathSegments.length < 1) {
+        continue;
+      }
+      String lastSegment = pathSegments[pathSegments.length - 1];
+      JsonNode currentNode = devfileNode;
+      // iterate until we reach last but one path segment,
+      // because it's impossible to change value of the current Json node,
+      // so to set value to Json node, you should be 1 level up and do put(key, value).
+      Iterator<String> pathSegmentsIterator =
```

Comment:
Why are you creating an exact copy of the `pathSegments` here?

Reply:
Also, would it work if you just really iterated to the second last element using `Stream.of(pathSegments).limit(pathSegments.length - 1).iterator()` and then handled the last segment? Maybe it would simplify the logic in the loop where we would not have to check if we're at the last segment or not.

Reply:
1) refactoring bug :) inititally it was doing copy with `length - 1`,
BUT
2) `length-1` is not works when you have just one segment provided, you still need at least one iteration, thats why it is designed in `do-while` cycle.

Reply:
UPD: also fo cases when it's like <array>.<item> we need the <item> to be in the iterator as well